### PR TITLE
New sample PDO contracts for managing policy-based access to OpenVINO ML models

### PR DESCRIPTION
**Pull request by prakashngit**
The PR introduces new sample PDO contracts for managing policy-based access to OpenVINO ML models. Files are located under inference-contract/. Please see inference-contract/README.md for overview of the new capabilities, as well instructions for testing the new samples.

### CODE REVIEW

File: inference-contract/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
... the implementation of a Private Data Objects contract family that implements a machine learning (ML) model based digital asset for performing policy-based inferencing operations using the underlying ML model. The protocol used for binding the asset to its use-policy is largely similar to the one used within the [Digital Asset Contract Family](../digital-asset-contract/README.md). The Token Object PDO contract specifies and checks the policies to be followed to use the ML model for the inferecing operation. The asset guardian service is implemented as a web service containing a frontend and backend portions. In this implementation, the ML Model is assumed to be an [OpenVINO based IR](https://docs.openvino.ai/2023.2/openvino_ir.html) model.  The backend of the guardian service in this case implements the OpenVINO model server capable of performing inferencing operations using the OpenVINO IR model. The frontend of the guardian service at a high level performs two functions: 1. Processes the capability package generated by the token object and generates the API for the backend. 2. Implements any input preprocessing and output postprocessing (collectively referred to as model scoring)
```
Comment (marcelamelara) [inline inference-contract/README.md:7]: Stylistic: Please wrap these lines at 80 chars. This will make it much easier to review sentence by sentence.

Comment (marcelamelara) [inline inference-contract/README.md:7]: A couple edit suggestions and comments: 

* The first sentence is a little tautological. 
* I would present the frontend of the guardian service before presenting the backend. From a user's perspective, they'll be interacting with that piece, not the backend.
* If you have time, I'd highly recommend adding a basic figure that shows the protocol as a visual aid, even if the protocol is similar to the one in the DA contracts. It'll help any user of the contracts get an idea of what's going on much faster than having to click through links.

```suggestion
This directory contains a Private Data Objects contract family that implements a machine learning (ML) model based digital asset for performing policy-based inferencing operations using the underlying ML model. The protocol used for binding the asset to its use-policy is largely similar to the one used within the [Digital Asset Contract Family](../digital-asset-contract/README.md). The Token Object PDO contract specifies and checks the policies to be followed to use the ML model for the inferecing operation. The asset guardian service is implemented as a web service containing a frontend and backend portions. In this implementation, the ML Model is assumed to be an [OpenVINO based IR](https://docs.openvino.ai/2023.2/openvino_ir.html) model.  The backend of the guardian service in this case implements the OpenVINO model server capable of performing inferencing operations using the OpenVINO IR model. The frontend of the guardian service at a high level performs two functions: 1. Processes the capability package generated by the token object and generates the API for the backend. 2. Implements any input preprocessing and output postprocessing (collectively referred to as model scoring)
```

File: inference-contract/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+as determined necessary by the original asset owner. 
+
+The token object, as well as the asset guardian service are deployed by the asset owner, as part of the deployment sequence. The asset owner in addition authors the model scoring script that is executed as part of the guardian service frontend. Note that the model owner may use the token object to control key parameters used by the scoring script, and thus ensuring tight binding between the token object contract and the model usage.
```
Comment (marcelamelara) [inline inference-contract/README.md:10]: Same 80 char line wrapping comment as above. And this is really where a figure would be helpful.

File: inference-contract/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+```bash
+mkdir -p imageclassification/model/1
+wget -P imageclassification/model/1 https://storage.openvinotoolkit.org/repositories/open_model_zoo/2022.1/models_bin/2/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.bin
+wget -P imageclassification/model/1 https://storage.openvinotoolkit.org/repositories/open_model_zoo/2022.1/models_bin/2/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.xml
+docker run -d -u $(id -u):$(id -g) -v $(pwd)/imageclassification/model/:/models/ -p 127.0.0.1:9000:9000 openvino/model_server:latest --model_path /models/ --model_name resnet --port 9000 --shape auto
+```
+
+The model server endpoint is `127.0.0.1:9000`. Please note that we want the model server to be accessible only from the Guardian Service front end. In our tests, we assume that the Guardian Service and the model server are hosted on the same machine. The token owner (in practice siutated on a remote node) submits inference capability requests to the Guardian Service. The Guardian Service verifies the capability request,  gets its processed via the model server backend, and sends the result back to the token owner.
```
Comment (marcelamelara) [inline inference-contract/README.md:29]:
```suggestion
The model server endpoint is `127.0.0.1:9000`. Please note that we want the model server to be accessible only from the Guardian Service front end. In our tests, we assume that the Guardian Service and the model server are hosted on the same machine. The token owner (in practice situated on a remote node) submits inference capability requests to the Guardian Service. The Guardian Service verifies the capability request,  gets it processed via the model server backend, and sends the result back to the token owner.
```

File: inference-contract/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+as determined necessary by the original asset owner. 
+
+The token object, as well as the asset guardian service are deployed by the asset owner, as part of the deployment sequence. The asset owner in addition authors the model scoring script that is executed as part of the guardian service frontend. Note that the model owner may use the token object to control key parameters used by the scoring script, and thus ensuring tight binding between the token object contract and the model usage.  
+
+## Deploying the OpenVINO Model Server Backend for tests ##
```
Comment (marcelamelara) [inline inference-contract/README.md:12]: Nit: make it clear up front that this is for testing.
```suggestion
## Test Deployment of the OpenVINO Model Server Backend ##
```

File: inference-contract/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+```
+
+The model server endpoint is `127.0.0.1:9000`. Please note that we want the model server to be accessible only from the Guardian Service front end. In our tests, we assume that the Guardian Service and the model server are hosted on the same machine. The token owner (in practice siutated on a remote node) submits inference capability requests to the Guardian Service. The Guardian Service verifies the capability request,  gets its processed via the model server backend, and sends the result back to the token owner.
+
+## Testing the inference contract ##
+
+The tests assume that PDO services and ledger are up and running. Use the variables `F_SERVICE_HOST`and `PDO_LEDGER)URL` in `script_test.sh` to respectively specify the IP address for the PDO services and the URL For the PDO ledger. The script test will locally deploy the guardian service frontend necessary for the test. Note that the guardian front end uses a local storage service, that gets deployed (as part of the test script) prior to starting the front end. The storage service is used as an efficient means for the token owner to send potentially large input images for the inferencing operations.
```
Comment (marcelamelara) [inline inference-contract/README.md:33]:
```suggestion
The tests assume that PDO services and ledger are up and running. Use the variables `F_SERVICE_HOST` and `PDO_LEDGER_URL` in `script_test.sh` to respectively specify the IP address for the PDO services and the URL for the PDO ledger. The script test will locally deploy the guardian service frontend necessary for the test. Note that the guardian frontend uses a local storage service, that gets deployed (as part of the test script) prior to starting the frontend. The storage service is used as an efficient means for the token owner to send potentially large input images for the inferencing operations.
```

**Review by marcelamelara — COMMENTED** [review]
README review only.

File: inference-contract/README.md (hunk 3 above)
Comment (g2flyer) [inline inference-contract/README.md:29]: Could be worth mentioning that port 9000 is defined in `etc/guardian_service.toml` (and hence could also be here and there changed if necessary)?

BTW: force-pushing during reviews is not really ideal; made me loose some (still relevant) comments ;-)  [put differently fixup commits are nicer although i have to admit as github (still?) doesn't not handle auto-squash automatically it requires a bit more complication during merge ....

File: inference-contract/README.md (hunk 5 above)
Comment (g2flyer) [inline inference-contract/README.md:33]: The script also assumes the model server is up (as described above)?

**Review by g2flyer — COMMENTED** [review]
Following some quick comments on README.

File: inference-contract/README.md (hunk 3 above)
Comment (cmickeyb) [reply inference-contract/README.md:29]: this is my fault. sort of.
i wanted more details in the commits.

Comment (g2flyer) [reply inference-contract/README.md:29]: > this is my fault. sort of.
> i wanted more details in the commits.

I guess you are referring to my BTW?  Thinking of it, i guess it's true that fixup makes it easy to push code changes, there doesn't seem to be an easy way (other than delaying it to the autosquash) of updating commit log messages ...

**Review by cmickeyb — CHANGES_REQUESTED** [review]
Please reformat the README with 80 or 100 character line limits.

**Review by cmickeyb — APPROVED** [review]

**Pull request merged**